hyperspy/io.py

```diff
@@ -848,3 +852,51 @@ def save(filename, signal, overwrite=None, **kwds):
                 signal.tmp_parameters.set_item('folder', file.parent)
                 signal.tmp_parameters.set_item('filename', file.stem)
                 signal.tmp_parameters.set_item('extension', extension)
+
+
+def add_file_load_metadata(signal_dict, reader):
+    mdata_dict = {
+        'operation': 'load',
+        'io_plugin': reader.__loader__.name,
+        'hyperspy_version': hs_version,
+        'timestamp': datetime.now().astimezone().isoformat()
+    }
+    if 'FileIO' not in signal_dict['metadata']['General']:
+        signal_dict['metadata']['General']['FileIO'] = {
+            '0': mdata_dict
+        }
+    else:
+        largest_index = max([
+            int(i.replace('Number_', '')) for i in
+            signal_dict['metadata']['General']['FileIO'].keys()])
+        signal_dict['metadata'][
+            'General']['FileIO'][str(largest_index + 1)] = mdata_dict
+
+    return signal_dict
+
+
+def add_file_save_metadata(signal, writer):
+    mdata_dict = {
+        'operation': 'save',
+        'io_plugin': writer.__loader__.name,
+        'hyperspy_version': hs_version,
+        'timestamp': datetime.now().astimezone().isoformat()
+    }
+    if 'General' not in signal.metadata or isinstance(signal.metadata, dict):
+        # if we don't even have basic metadata as a DictionaryTreeBrowser,
+        # just return as-is (shouldn't happen, but this came up in tests of
+        # test_nexus_hdf.test_save_metadata_as_dict())
+        return signal
+    if 'FileIO' not in signal.metadata.General:
+        signal.metadata.General.add_dictionary = {
+            '0': mdata_dict
+        }
+    else:
+        largest_index = \
+            max([int(i.replace('Number_', '')) for i in
+                 signal.metadata.General.FileIO.keys()])
+        signal.metadata.General.FileIO.add_dictionary({
+            f'{largest_index + 1}': mdata_dict
+        })
+
+    return signal
```

Comment:
These two functions could be merged by adding an argument, which specifies if the key is `'save'` or `'load'`.